// music-metadata/lib/id3v1/ID3v1TagMap.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var GenericTagMapper_1 = require("../common/GenericTagMapper");
/**
 * ID3v1 tag mappings
 */
var id3v1TagMap = {
    title: 'title',
    artist: 'artist',
    album: 'album',
    year: 'year',
    comment: 'comment',
    track: 'track',
    genre: 'genre'
};
var ID3v1TagMapper = /** @class */ (function (_super) {
    __extends(ID3v1TagMapper, _super);
    function ID3v1TagMapper() {
        return _super.call(this, ['ID3v1.1'], id3v1TagMap) || this;
    }
    return ID3v1TagMapper;
}(GenericTagMapper_1.CommonTagMapper));
exports.ID3v1TagMapper = ID3v1TagMapper;
