// music-metadata/lib/vorbis/VorbisParser.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var Vorbis = require("./Vorbis");
var Token = require("token-types");
var _debug = require("debug");
var debug = _debug("music-metadata:parser:Ogg/Vorbis1");
/**
 * Vorbis 1 Parser.
 * Used by OggParser
 */
var VorbisParser = /** @class */ (function () {
    function VorbisParser(options) {
        this.options = options;
        this.format = {};
        this.tags = [];
        this.pageSegments = [];
    }
    /**
     * Vorbis 1 parser
     * @param header Ogg Page Header
     * @param pageData Page data
     */
    VorbisParser.prototype.parsePage = function (header, pageData) {
        if (header.headerType.firstPage) {
            this.parseFirstPage(header, pageData);
        }
        else {
            if (header.headerType.continued) {
                if (this.pageSegments.length === 0) {
                    throw new Error("Cannot continue on previous page");
                }
                this.pageSegments.push(pageData);
            }
            if (header.headerType.lastPage || !header.headerType.continued) {
                // Flush page segments
                if (this.pageSegments.length > 0) {
                    var fullPage = Buffer.concat(this.pageSegments);
                    this.parseFullPage(fullPage);
                }
                // Reset page segments
                this.pageSegments = header.headerType.lastPage ? [] : [pageData];
            }
        }
        if (header.headerType.lastPage) {
            this.calculateDuration(header);
        }
    };
    VorbisParser.prototype.flush = function () {
        this.parseFullPage(Buffer.concat(this.pageSegments));
    };
    VorbisParser.prototype.getMetadata = function () {
        return {
            format: this.format,
            native: {
                vorbis: this.tags
            }
        };
    };
    /**
     * Parse first Ogg/Vorbis page
     * @param {IPageHeader} header
     * @param {Buffer} pageData
     */
    VorbisParser.prototype.parseFirstPage = function (header, pageData) {
        debug("Parse first page");
        // Parse  Vorbis common header
        var commonHeader = Vorbis.CommonHeader.get(pageData, 0);
        if (commonHeader.vorbis !== 'vorbis')
            throw new Error('Metadata does not look like Vorbis');
        this.format.dataformat = "Ogg/Vorbis I";
        if (commonHeader.packetType === 1) {
            var idHeader = Vorbis.IdentificationHeader.get(pageData, Vorbis.CommonHeader.len);
            this.format.sampleRate = idHeader.sampleRate;
            this.format.bitrate = idHeader.bitrateNominal;
            this.format.numberOfChannels = idHeader.channelMode;
            debug("sample-rate=%s[hz], bitrate=%s[b/s], channel-mode=%s", idHeader.sampleRate, idHeader.bitrateNominal, idHeader.channelMode);
        }
        else
            throw new Error('First Ogg page should be type 1: the identification header');
    };
    VorbisParser.prototype.parseFullPage = function (pageData) {
        // New page
        var commonHeader = Vorbis.CommonHeader.get(pageData, 0);
        debug("Parse full page: type=%s, byteLength=%s", commonHeader.packetType, pageData.byteLength);
        switch (commonHeader.packetType) {
            case 3: //  type 3: comment header
                return this.parseUserCommentList(pageData, Vorbis.CommonHeader.len);
            case 1: // type 1: the identification header
            case 5: // type 5: setup header type
                break; // ignore
        }
    };
    VorbisParser.prototype.calculateDuration = function (header) {
        if (this.format.sampleRate && header.absoluteGranulePosition >= 0) {
            // Calculate duration
            this.format.numberOfSamples = header.absoluteGranulePosition;
            this.format.duration = this.format.numberOfSamples / this.format.sampleRate;
        }
    };
    /**
     * Ref: https://xiph.org/vorbis/doc/Vorbis_I_spec.html#x1-840005.2
     * @returns {Promise<number>}
     */
    VorbisParser.prototype.parseUserCommentList = function (pageData, offset) {
        var strLen = Token.UINT32_LE.get(pageData, offset);
        offset += 4;
        var vendorString = new Token.StringType(strLen, 'utf-8').get(pageData, offset);
        offset += strLen;
        var userCommentListLength = Token.UINT32_LE.get(pageData, offset);
        offset += 4;
        while (userCommentListLength-- > 0) {
            offset += this.parseUserComment(pageData, offset);
        }
    };
    VorbisParser.prototype.parseUserComment = function (pageData, offset) {
        var strLen = Token.UINT32_LE.get(pageData, offset);
        var v = new Token.StringType(strLen, 'utf-8').get(pageData, offset + 4);
        var idx = v.indexOf('=');
        var key = v.slice(0, idx).toUpperCase();
        var value = v.slice(idx + 1);
        if (key === 'METADATA_BLOCK_PICTURE') {
            value = this.options.skipCovers ? null : Vorbis.VorbisPictureToken.fromBase64(value);
        }
        if (value !== null) {
            debug("Push tag: id=%s, value=%s", key, value);
            this.tags.push({ id: key, value: value });
        }
        return Token.UINT32_LE.len + strLen;
    };
    return VorbisParser;
}());
exports.VorbisParser = VorbisParser;
