// music-metadata/lib/aiff/AiffParser.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var strtok3 = require("strtok3");
var Token = require("token-types");
var Chunk = require("./Chunk");
var stream_1 = require("stream");
var ID3v2Parser_1 = require("../id3v2/ID3v2Parser");
var FourCC_1 = require("../common/FourCC");
/**
 * AIFF - Audio Interchange File Format
 *
 * Ref:
 *  http://www.onicos.com/staff/iz/formats/aiff.html
 *  http://muratnkonar.com/aiff/index.html
 */
var AIFFParser = /** @class */ (function () {
    function AIFFParser() {
        this.metadata = {
            format: {
                dataformat: "AIFF",
                tagTypes: []
            },
            native: {}
        };
    }
    AIFFParser.prototype.parse = function (tokenizer, options) {
        var _this = this;
        this.tokenizer = tokenizer;
        this.options = options;
        return this.tokenizer.readToken(Chunk.Header)
            .then(function (header) {
            if (header.chunkID !== 'FORM')
                throw new Error("Invalid Chunk-ID, expected 'FORM'"); // Not AIFF format
            return _this.tokenizer.readToken(FourCC_1.FourCcToken).then(function (type) {
                _this.metadata.format.dataformat = type;
            }).then(function () {
                return _this.readChunk().then(function () { return _this.metadata; });
            });
        });
    };
    AIFFParser.prototype.readChunk = function () {
        var _this = this;
        return this.tokenizer.readToken(Chunk.Header)
            .then(function (header) {
            switch (header.chunkID) {
                case 'COMM': // The Common Chunk
                    return _this.tokenizer.readToken(new Chunk.Common(header))
                        .then(function (common) {
                        _this.metadata.format.bitsPerSample = common.sampleSize;
                        _this.metadata.format.sampleRate = common.sampleRate;
                        _this.metadata.format.numberOfChannels = common.numChannels;
                        _this.metadata.format.numberOfSamples = common.numSampleFrames;
                        _this.metadata.format.duration = _this.metadata.format.numberOfSamples / _this.metadata.format.sampleRate;
                    });
                case 'ID3 ': // ID3-meta-data
                    return _this.tokenizer.readToken(new Token.BufferType(header.size))
                        .then(function (id3_data) {
                        var id3stream = new ID3Stream(id3_data);
                        return strtok3.fromStream(id3stream).then(function (rst) {
                            return ID3v2Parser_1.ID3v2Parser.getInstance().parse(_this.metadata, rst, _this.options);
                        });
                    });
                case 'SSND': // Sound Data Chunk
                default:
                    return _this.tokenizer.ignore(header.size);
            }
        }).then(function () { return _this.readChunk(); }).catch(function (err) {
            if (err.message !== strtok3.endOfFile) {
                throw err;
            }
        });
    };
    return AIFFParser;
}());
exports.AIFFParser = AIFFParser;
var ID3Stream = /** @class */ (function (_super) {
    __extends(ID3Stream, _super);
    function ID3Stream(buf) {
        var _this = _super.call(this) || this;
        _this.buf = buf;
        return _this;
    }
    ID3Stream.prototype._read = function () {
        this.push(this.buf);
        this.push(null); // push the EOF-signaling `null` chunk
    };
    return ID3Stream;
}(stream_1.Readable));
